Introduce accessible prompter for screen readers (preview)

BagToad opened the pull request:
## Description

Implement `charmbracelet/huh` in accessible mode as an accessible prompter.


## Acceptance Criteria


> **When** I run `gh environment`
> **Then** I see the `GH_SCREENREADER_FRIENDLY` env var documented and marked as experimental
> 
> **Given** I have `GH_SCREENREADER_FRIENDLY` set to any value other than `0` or `false`
> **When** I run any `gh` command requiring prompting
> **Then** the prompting experience provides a basic level of screenreader friendliness

## Demo

<details><summary>Details</summary>
<p>

```
Demonstrating Confirmation
Are you sure?

Choose [y/N]: y
Chose: Yes

Confirmation: true
Demonstrating Text Input
Favorite meal? (Breakfast)

Input: lunch
Input: lunch
            
Favorite meal: lunch
Demonstrating Single Select
Favorite cuisine?
1. Italian
2. Greek
3. Indian
4. Japanese
5. American

Choose: 5
Chose: American
               
Favorite cuisine: American
Demonstrating Multi Select
Favorite cuisines?
1.   Italian
2.   Greek
3.   Indian
4.   Japanese
5.   American

Select up to 5 options. 0 to continue.
Select: 5
Selected: American

Favorite cuisines?
1.   Italian
2.   Greek
3.   Indian
4.   Japanese
5. ✓ American

Select up to 5 options. 0 to continue.
Select: 3
Selected: Indian

Favorite cuisines?
1.   Italian
2.   Greek
3. ✓ Indian
4.   Japanese
5. ✓ American

Select up to 5 options. 0 to continue.
Select: 1
Selected: Italian

Favorite cuisines?
1. ✓ Italian
2.   Greek
3. ✓ Indian
4.   Japanese
5. ✓ American

Select up to 5 options. 0 to continue.
Select: 0
Selected: Italian, Indian, American
                                   
Favorite cuisine: Italian
Favorite cuisine: Indian
Favorite cuisine: American
Favorite meal: lunch
Demonstrating Password Input (unmasked)
Safe word?

Input: stuff
Input: stuff
            
Safe word: stuff
Confirmation: true
Demonstrating Auth Token (can't be blank)
Paste your authentication token:

Input: blalbla
Input: blalbla
              
Auth token: blalbla
Demonstrating Deletion Confirmation
Type "delete-me" to confirm deletion

Input: delete-me
Input: delete-me
                
Item deleted
Demonstrating Hostname (must be valid hostname)
Hostname:

Input: github.com
Input: github.com
                 
Hostname: github.com
Demonstrating Markdown Editor with blanks allowed
Edit your text:
1. Open Editor: vim
2. Skip

Choose: 2
Chose: Skip
           
Edited text: 
Demonstrating Markdown Editor with blanks disallowed
Edit your text:
1. Open Editor: vim

Choose: 1
Chose: Open Editor: vim
                       
Edited text: Initial text
```

</p>
</details> 


## Notes

- I have identified a difference in behavior between this prompter and the current one, described in https://github.com/charmbracelet/huh/issues/615. I left a commented unit test in which can be used if this behavior is added to `huh` in the future.
- After discussing with @williammartin and @andyfeller, we decided to rename the env var to `GH_ACCESSIBLE_PROMPTER` for now.

## Review

Consider checking out and running #10745 for a quick demo :)

BagToad committed:
feat(prompter): add accessible prompter support

BagToad committed:
feat(md prompter): md prompt respects blankAllowed

Accessible prompter now respects blankAllowed and
will not prompt for "skip" if blankAllowed is false.

BagToad committed:
fix(md prompter): accessible prompt allows blank

Allow the accessible markdownEditor prompt to be blank when the blank
comes from the result of an interactive session with an editor, even when
blankAllowed is false.

This behavior aligns the accessible prompter with the behavior of the
current standard prompter.

BagToad committed:
test(prompter): remove t.parallel calls

t.Parallel() cannot be used when env vars are being set.

andyfeller commented on internal/prompter/prompter.go:
```diff
@@ -27,15 +30,232 @@ type Prompter interface {
-		editorCmd: editorCmd,
+	accessiblePrompterValue := os.Getenv("GH_SCREENREADER_FRIENDLY")
+	switch accessiblePrompterValue {
+	case "", "false", "0":
```
```suggestion
	case "", "false", "0", "no":
```

andyfeller replied:
I realize our AC might not have touched on this, but I want to encourage adding `"no"` as a non-truthy option as we do in [other places](https://github.com/search?q=org%3Acli+%22%5C%22no%5C%22%22&type=code)

williammartin commented on internal/prompter/prompter_test.go:
```diff
@@ -0,0 +1,58 @@
+	"github.com/stretchr/testify/assert"
+)
+
+func TestNewReturnsAccessiblePrompter(t *testing.T) {
```
**Suggestion**

I think generally we might want to refer to this thing as `screenreaderFriendlyPrompter` rather than `accessiblePrompter`. It's a clearer link to the env var, and accessibility is a big scope.

BagToad replied:
Discussed with @williammartin  - we decided to go with `SpeechSynthesizerFriendlyPrompter`

BagToad replied:
Done in 94bbd26aab390c3f8929b040ec229484b9685bd3 and e42af358392fae34eab89053c76e286a10653815

williammartin commented on internal/prompter/prompter_test.go:
```diff
@@ -0,0 +1,58 @@
+
+		p := New(editorCmd, stdin, stdout, stderr)
+
+		assert.IsType(t, &huhPrompter{}, p, "expected huhPrompter to be returned")
```
**Suggestion**

I think no assertion after this makes sense if this one fails, so prefer `require.IsType`.

BagToad replied:
done in 0d7fd36f11ead4d35d09b51a00ba70ac54d5131d

shindere commented:
William Martin (2025/04/02 04:37 -0700):
> @williammartin commented on this pull request.
> I think generally we might want to refer to this thing as
> `screenreaderFriendlyPrompter` rather than `accessiblePrompter`. It's
> a clearer link to the env var, and accessibility is a big scope.

Well, if I may say, I find both namigs a bit vague in the sense that
they say what they would like to achieve rather than what they are
effectively and precisely doing.

I was curious about what this PR does but I couln't see a description
for it? Did I miss something maybe?

williammartin commented on internal/prompter/prompter.go:
```diff
@@ -27,15 +30,232 @@ type Prompter interface {
+}
+
+// IsAccessible returns true if the huhPrompter was created in accessible mode.
+func (p *huhPrompter) IsAccessible() bool {
```
**Question**

What is this method providing us? There is no inaccessible prompting for `huhPrompter`. The only reason we had a toggle in the spike was to compare versions.

BagToad replied:
@williammartin it is only providing us the ability to inspect the prompter in unit tests for assurance that we do indeed get an accessible prompter.

```Go
assert.Equal(t, p.(*huhPrompter).IsAccessible(), true, "expected huhPrompter to be accessible")
```

- Do you have thoughts on whether this testing is valuable or not?
- Do you have thoughts on another way to gain this assurance, if you agree that it's valuable to test this?

BagToad replied:
Discussed and removed this code in favour of behavioral focused testing.

williammartin commented:
> I was curious about what this PR does but I couln't see a description for it? Did I miss something maybe?

Firstly, this is in a draft state, which is why there is no description. You also haven't missed anything on our public issue tracker, but I'm glad you dropped by! One of our findings using screenreaders (and indicated by another screenreader user that consults with GitHub) was that the prompting experience right can be challenging to use because of the way it redraws the screen.

The goal here is to ship a prompting experience behind the `GH_SCREENREADER_FRIENDLY` environment variable that we can get feedback, and iterate on. We'll make sure there is a public place for discussion.

> Well, if I may say, I find both namings a bit vague in the sense that they say what they would like to achieve rather than what they are effectively and precisely doing.

Thanks for the feedback. For the moment I think that this is fine, the goal is that this prompting experience is more friendly for screenreaders. The preciseness will be in the concrete implementation as opposed to the behavioural specification of the tests.

shindere commented:
William Martin (2025/04/02 04:59 -0700):
> Firstly, this is in a draft state, which is why there is no
> description.

ACK, thanks!

> You also haven't missed anything on our public issue tracker, but I'm
> glad you dropped by! One of our findings using screenreaders (and
> indicated by another screenreader user that consults with GitHub) was
> that the prompting experience right can be challenging to use because
> of the way it redraws the screen.

I see. As a sidenote I'd definitely consider consulting.

I guess what puzzles me is the vagueness of all this. I mean, I do
understand that it's too early to be precise, but I also would like to
emphasize that, to me, this very lack of precision itself may be a
source of future issues for you GitHub guys.

Indeed, it feels to me here as if it is believed that there is one way
to make something `screenreader friendly`. But I fear the reality is
more complex: depending on whether you use speech synthesis or a braille
display, also depending on the environemnt you work with (where
environemnt means OS, framework and screenreader), your needs in terms
of accessibility and even your definitionof what accessibility is may
vary. Evenmore so as there is nothing here like WCAG to guide design.

Not to say what's currently being done is wrong, it certainly isn't. But
you man want to keep all this in mind, in case it was not already
obvious, in which case I sincerely apologize.

williammartin commented:
> Even more so as there is nothing here like WCAG to guide design.

Haha yes, we're familiar with the challenge of no WCAG, as we, GitHub, and Microsoft have had to learn what it means to have different forms of accessibility in the terminal. I'm sure we still have a lot to learn.

> Indeed, it feels to me here as if it is believed that there is one way to make something `screenreader friendly`. But I fear the reality is more complex: depending on whether you use speech synthesis or a braille display, also depending on the environment you work with (where environment means OS, framework and screenreader), your needs in terms
of accessibility and even your definition of what accessibility is may vary.

This is really fantastic insight thank you. I suppose in this case we are referring to a speech synthesis friendly prompting experience. 

For example, a single select would read like:

```
Favorite cuisine?
1. Italian
2. Greek
3. Indian
4. Japanese
5. American

Choose: 2
Chose: Greek
```

A multi-select would read like:

```
Favorite cuisines?
1.   Italian
2.   Greek
3.   Indian
4.   Japanese
5.   American

Select up to 5 options. 0 to continue.
Select: 1
Selected: Italian

Favorite cuisines?
1. ✓ Italian
2.   Greek
3.   Indian
4.   Japanese
5.   American

Select up to 5 options. 0 to continue.
Select: 3
Selected: Indian

Favorite cuisines?
1. ✓ Italian
2.   Greek
3. ✓ Indian
4.   Japanese
5.   American

Select up to 5 options. 0 to continue.
Select: 0
Selected: Italian, Indian
```

An input would read like:

```
Favorite meal?

Input: pizza
Input: pizza
```

And so on. I'm sure there's a lot of room for improvement in the exact styles and text, but hopefully you understand the goal we're going for here in comparison to our current prompting experience (you can run `gh auth login` with no arguments for an example of a command with multiple prompts).

Do you think this would be best referred to as a Speech Synthesis Friendly Prompter?

Braille displays are definitely something we are less familiar with supporting. In that regard:

> As a sidenote I'd definitely consider consulting.

I'll pass this information on to the accessibility team.

shindere commented:
William Martin (2025/04/02 05:22 -0700):
> This is really fantastic insight thank you. I suppose in this case we
> are referring to a speech synthesis friendly prompting experience.

Okay.

> Do you think this would be best referred to as a Speech Synthesis Friendly Prompter?

Possibly, yes.

Also: I completely understand that you'd like to experiment things, get
feedback etc. without disturbing existing workflows and I find this both
very kind and mindful so what I will say now only applies to an ideal,
ultimate state of things:

I know that at least as a user of braille, I would not feel comfortable
with the idea of having to use a different prompter jsut because I use
braille. One reason is technical, it is known that such things are less
used, tested and then less maintained. The other reason is more social:
I find it a bit discriminating if I have to use another, dedicated way
and I feel way more included if I can use what others, because the way
it has been designed is universalenough for that thing to work for me,
too.

> And so on. I'm sure there's a lot of room for improvement in the exact
> styles and text, but hopefully you understand the goal we're going for
> here in comparison to our current prompting experience (you can run
> `gh auth login` with no arguments for an example of a command with
> multiple prompts).

Done. On the first screen I see a question mark at the very beginning of
the message, which surprises me.

Apart frm that I find it works pretty well, especially the cursor is
where the relevant information is, which, IMO, is nice.

I guess what I would personnaly like for this prompter, as much as a
geekthan as a blind user, is configurability. See mutt and its
mitt_index variable for instance which can include %-prefixed characters
that let you build your line exactly like you want. Souser mays want to
remove the > sign to gain space. Other with a longer display may want to
add the number of the current option and the number of total options,
like 1/3, 2/3, 3/3. Some may want to have this at the beginning and
others at the end.

And finally that would benefit to all and there is not even a need to
"claim" accessibility, it's just a by-product of clever design.

(FWIW mutt DOES have a braile_friendly variable.)


> 
> Braille displays are definitely something we are less familiar with supporting. In that regard:
> 
> > As a sidenote I'd definitely consider consulting.
> 
> I'll pass this information on to the accessibility team.

Thanks!

williammartin commented:
All very useful feedback thank you. We'll need to figure out a way to distil this conversation into some document, rather than just living on a PR.

> Also: I completely understand that you'd like to experiment things, get feedback etc. without disturbing existing workflows and I find this both very kind and mindful so what I will say now only applies to an ideal, ultimate state of things.

At the end of the day, we're open to wherever our users lead us, even more so on this topic where we don't have the lived experience. As you say, in an ideal world, the out of the box experience is shared and accessible by default, and we have been doing work on various fronts towards that. Your surprise over the starting question mark is something we could look at addressing.

> I guess what I would personally like for this prompter, as much as a geek than as a blind user, is configurability.

Makes sense to me. One of the challenges we have at the moment is that the current prompting library is deprecated. If you could wave a magic wand and have configurability for the prompts, what options would you like? Would it be useful if I produced a program that demonstrated each different type of prompt?

Food for thought on the configurability side, is where the boundary between "different prompting experience" and "configurable prompting experience is". I definitely appreciate your shared concerns though, will need to reflect on this. Currently this new prompting experience is going to live behind an experimental configuration so that we can make sure to get this right. If it turns out that "right" is getting rid of it, that's good to know.

> Apart from that I find it works pretty well, especially the cursor is where the relevant information is, which, IMO, is nice.

That's nice to know. Just to be clear, were you using the braille reader, or speech synthesis?

One issue we discovered was that the selects in screen synthesis would often announce "blank" when moving between options.

shindere commented:
William Martin (2025/04/02 06:15 -0700):
> > I guess what I would personally like for this prompter, as much as a geek than as a blind user, is configurability.
> 
> Makes sense to me. One of the challenges we have at the moment is that
> the current prompting library is deprecated.

Oops, sorry to hear that.

> If you could wave a magic wand and have configurability for the
> prompts, what options would you like?

I fear I didn't understand your quesiton, sorry about that.

What I did notice is the text saying to press arrow keys. That I think
is something one may want to be able to suppress because, once you know,
I think it will be perceived more as noise than as help by users, here
especially by users of speech sysnthesizers as I think it's less easy
for them to skip text than it is is braille (to be confirmed).

> Would it be useful if I produced a program that demonstrated each
> different type of prompt?

I am worried about making you lose time and energy: I am unsure we need
such a program, I mean perhaps there are ways to achieve the same goal
without writing a dedicated programs, like providing a subset of
commands that will cover all the types of prompts.

But of course a program may also have its own advantages if the purpose
is to then write automated tests.

> Food for thought on the configurability side, is where the boundary
> between "different prompting experience" and "configurable prompting
> experience is".

Out of the top of my head and without having thought about it for a long
time, I have a vague feeling that the difference is that configuring is
something you can do if you want (a possibilit which is offered to you)
but that you do not _have to do_ because the thingwould otherwise be
unusable. But this is a very, very personal idea.

> I definitely appreciate your shared concerns though, will need to
> reflect on this. Currently this new prompting experience is going to
> live behind an experimental configuration so that we can make sure to
> get this right. If it turns out that "right" is getting rid of it,
> that's good to know.

To avoid biases as much as possible, you may want to proactively reach
out to potentially concerned ussers early, that would avoid you doing
many steps in one direction before noticing this directiion is notthat
good.

For instance the developers of the NVDA Windows screen reader use GitHub
and several of them are visually impaired. You may want to contact them.
On the braille side, you cna reach out to many people with a geek /
sysadmin / developer background through the ***@***.***,
***@***.*** and ***@***.*** mailing lists.

Beyond the exchanges you may have there about the gh CLI, I think that,
just by subscribing to these mailing lists you can learn quite a bunch
of things about what is difficult.



> 
> > Apart from that I find it works pretty well, especially the cursor is where the relevant information is, which, IMO, is nice.
> 
> That's nice to know.

(I think it was already me who reported the problem of the cursor not
being on the selected item...)

> Just to be clear, were you using the braille reader, or speech
> synthesis?

Braille display.
 
> One issue we discovered was that the selects in screen synthesis would
> often announce "blank" when moving between options.

Okay but that is something I'm pretty sure has more to do with the way
the screenreader is configured and IMHO it's not your responsibility to
compensate for misconfigured tools (without being too extremist about
that, but still).

BagToad committed:
feat(prompter): include `no` as false-y value

Co-authored-by: Andy Feller <andyfeller@github.com>

BagToad committed:
test(prompter): `go-expect` based prompter tests

BagToad committed:
fix(prompter): rename huhprompter

BagToad committed:
test(prompter): replace assert with require

BagToad committed:
tests(prompter): rename huhprompter

BagToad committed:
test(prompter): remove needless variable declaration

BagToad committed:
test(prompter): skip vt10x tests on Windows

BagToad committed:
test(prompter): move to external package

BagToad committed:
fix(linter): linter errors

BagToad committed:
docs(prompter): doc prompter interface

BagToad committed:
refactor(prompter): rename env var for speech synthesizer friendly prompter

BagToad committed:
test(prompter): more tests for bad input

BagToad committed:
fix(prompter): input returns default when blank

BagToad commented:
related https://github.com/charmbracelet/huh/issues/615

BagToad committed:
fix(prompter): notes about Confirm default

BagToad committed:
fix(prompter): remove impossible condition

BagToad committed:
refactor(prompter): less magic strings

BagToad committed:
doc(prompter): clarify comments

BagToad committed:
doc(envs): speech synthesis prompter

BagToad committed:
test(prompter): fix race conditions

andyfeller commented on internal/prompter/prompter.go:
```diff
@@ -13,27 +15,262 @@ import (
-		stdout:    stdout,
-		stderr:    stderr,
-		editorCmd: editorCmd,
+	accessiblePrompterValue := os.Getenv("GH_SPEECH_SYNTHESIZER_FRIENDLY_PROMPTER")
+	switch accessiblePrompterValue {
+	case "", "false", "0", "no":
+		return &surveyPrompter{
+			prompter:  ghPrompter.New(stdin, stdout, stderr),
+			stdin:     stdin,
+			stdout:    stdout,
+			stderr:    stderr,
+			editorCmd: editorCmd,
+		}
+	default:
+		return &speechSynthesizerFriendlyPrompter{
+			stdin:     stdin,
+			stdout:    stdout,
+			stderr:    stderr,
+			editorCmd: editorCmd,
+		}
+	}
+}
```
**thought:** I realize the AC didn't call out the need for a `gh config` configuration setting to pair with the environment variable. 🤔  We should see about creating a follow up issue to address that.

When that happens, this will become an **issue** because `os.Getenv` will need to change to [`os.LookupEnv`](https://pkg.go.dev/os#LookupEnv) so the code doesn't interpret `""` from no env var being set as explicitly disabling the feature.

BagToad replied:
Should we just switch to `os.LookupEnv` in this PR then?

BagToad replied:
@andyfeller I took a stab at it in d8d3874778038fb5de581b7ed93b4e9bbaa4f185

I'll resolve this, but please unresolve it if you notice anything off or have more feedback about it ❤️

andyfeller commented on internal/prompter/prompter.go:
```diff
@@ -13,27 +15,262 @@ import (
-		stdout:    stdout,
-		stderr:    stderr,
-		editorCmd: editorCmd,
+	accessiblePrompterValue := os.Getenv("GH_SPEECH_SYNTHESIZER_FRIENDLY_PROMPTER")
```
**question:** Is this more confusing than the original proposal?

**praise:** I do think the env var should be narrowed down to the prompter.

**thought:** Shoot the difference with `GH_ACCESSIBLE_PROMPTER`, similar to `GH_ACCESSIBLE_COLORS`?

andyfeller replied:
I recognize and accept the original `GH_SCREENREADER_FRIENDLY` env var name was a little broad and vague when looked through the lens of this PR alone.  There are other features being discussed that were planned to piggy back on this, making it a larger setting to flip.

That said, I would like to simplify this to `GH_ACCESSIBLE_PROMPTER` for lack of a simpler env var if we're scoping this down.

Once we public preview changes and can talk with multiple users with lived experiences, we can lean into feedback and make changes accordingly.

BagToad replied:
Note: I spoke to @andyfeller about this already but wanted to write here more durably that I agree with changing it to align with the colors env var. I'll also note that I don't have too strong of an opinion here anymore, and really just would like a decision for the purpose of this PR.

@williammartin do you have any feelings on this that you want to float? I want to make sure we're all on the same page before I go make these changes again 😁

williammartin replied:
The change here was in response to Sebastian's feedback that using the words "screenreader friendly" was vague because it wasn't clear whether we were targeting speech synthesisers or braille readers.

If we did choose `ACCESSIBLE_PROMPTER` it probably has some of the same issues, and additionally wouldn't make sense for adjusting the spinner behaviour.

That said, I would be fine with any combination of:
 * `GH_SCREENREADER_FRIENDLY` for both
 * `GH_SPEECH_SYNTHESIZER_FRIENDLY_PROMPTER` and `GH_SPINNER=<val>` (or something else targeting spinner, like `LOADER`)
 * `GH_ACCESSIBLE_PROMPTER` and `GH_LOADER=<val>`

In all cases I anticipate that the env var(s) are marked as experimental, potential to be removed, and with clear explanations.

andyfeller replied:
Thank you for that insight! 🙇 ✨ Let me see about touching on a few points here.

### Problem with effect versus outcome

> The change here was in response to Sebastian's feedback that using the words "screenreader friendly" was vague because it wasn't clear whether we were targeting speech synthesizers or braille readers.

I really appreciate this insight because I definitely didn't consider it previously but in retrospect makes absolute sense.

My main concern with `GH_SPEECH_SYNTHESIZER_FRIENDLY_PROMPTER` is it based on some external result we cannot control and not what `gh` will do when it is turned on.  Additionally, these prompter changes should benefit both speech synthesizers and braille readers.

Perhaps `GH_ACCESSIBLE_PROMPTER` lacks language to explain that our user's interactive prompting experience will be flat or lacking any robust, cursor-redraw experience.

Outside of something like `GH_FLAT_PROMPTER` to describe the effect, `GH_ACCESSIBLE_PROMPTER` is as descriptive I can imagine. 🤔 

### Comparing this against other accessible CLI efforts

> * `GH_SPINNER=<val>` (or something else targeting spinner, like `LOADER`)

One insight that @BagToad shared previously was [`gcloud` accessibility features](https://cloud.google.com/sdk/docs/enabling-accessibility-features), which uses a broader property for toggling multiple experiences:

> In order to provide a more streamlined screen reader experience, the gcloud command-line tool comes with an `accessibility/screen_reader` property. When this property is set to true, the following behaviour is enabled:
> 
> - **Status trackers instead of unicode spinners**: The phrase 'working' is displayed on stderr while gcloud is performing tasks.
> - **Percentage progress bars**: Progress is displayed as a percentage, outputted to stderr.
> - **Flattened tables**: Boxed tables are the default output of many list commands. Instead of the queried resources being displayed in tables drawn in Unicode, results are rendered as a flattened list of items. Also, consider using the `--format` flag to define your own format.
> 
> To enable these accessibility features, run:
> 
> ```shell
> gcloud config set accessibility/screen_reader true
> ```

I only mention this to demonstrate inspiration from other CLIs and their approaches to accessibility approaches.

Part of our earlier discussions with the public preview was deferring more fine grained configuration / environment controls until we did user research and feedback post-public preview.

williammartin replied:
> Additionally, these prompter changes should benefit both speech synthesizers and braille readers.

That's fair, though at least in Sebastian's case using a braille reader, he thought the current prompting experience was quite good, and he would not like to use a different prompter. I'm still mulling this over, I think we should be receptive to the idea that this entire alternate prompter is not an experience we want in the name of accessibility (but perhaps for `TERM=dumb`).

Maybe `GH_MINIMAL_PROMPTER`? `GH_SIMPLE_PROMPTER`? `GH_BASIC_PROMPTER`?

> Part of our earlier discussions with the public preview was deferring more fine grained configuration / environment controls until we did user research and feedback post-public preview.

Makes sense to me.

andyfeller replied:
In lieu of strong user research after public preview, let's stick to `GH_ACCESSIBLE_PROMPTER` for now especially as this will align with other config / env var settings like `GH_ACCESSIBLE_COLORS`.

As mentioned in the discussion, using terms like _simple_ or _basic_ might be charged for some users more than others.  Again, we aren't even exposing the rich `huh` experience as part of public preview, which I think changes all this a bit more but let's save that for later too.

BagToad replied:
renamed in 19387b84187bef5edf437f4a46399843979e955e

williammartin replied:
> Again, we aren't even exposing the rich huh experience as part of public preview, which I think changes all this a bit more but let's save that for later too.

I don't know what the relevance is of this statement, but I'm fine with the rest of it.

BagToad committed:
test(prompter): add basic survey prompter test

andyfeller commented on internal/prompter/prompter.go:
```diff
@@ -13,27 +15,262 @@ import (
+
+func (p *speechSynthesizerFriendlyPrompter) Password(prompt string) (string, error) {
+	var result string
+	form := p.newForm(
+		huh.NewGroup(
+			huh.NewInput().
+				Title(prompt).
+				Value(&result),
+			// This doesn't have any effect in accessible mode.
+			// EchoMode(huh.EchoModePassword),
+		),
+	)
```
**nit:** I like the comment; maybe condense it as we don't know if support will be added

```suggestion
	// EchoMode(huh.EchoModePassword) doesn't have any effect in accessible mode.
	form := p.newForm(
		huh.NewGroup(
			huh.NewInput().
				Title(prompt).
				Value(&result),
		),
	)
```

andyfeller commented on internal/prompter/prompter.go:
```diff
@@ -13,27 +15,262 @@ import (
+}
+
+func (p *speechSynthesizerFriendlyPrompter) Confirm(prompt string, defaultValue bool) (bool, error) {
+	// This is currently an inneffectual assignment because the value is
```
**nit:** minor typo

```suggestion
	// This is currently an ineffectual assignment because the value is
```

andyfeller commented on internal/prompter/prompter.go:
```diff
@@ -13,27 +15,262 @@ import (
+
+func (p *speechSynthesizerFriendlyPrompter) ConfirmDeletion(requiredValue string) error {
+	var result string
+	form := p.newForm(
+		huh.NewGroup(
+			huh.NewInput().
+				Title(fmt.Sprintf("Type %q to confirm deletion", requiredValue)).
+				Validate(func(input string) error {
+					if input != requiredValue {
+						return fmt.Errorf("You entered: %q", input)
+					}
+					return nil
+				}).
+				Value(&result),
+			// This doesn't have any effect in accessible mode.
+			// EchoMode(huh.EchoModePassword),
+		),
+	)
```
**nit:** similar to above

```suggestion
	// EchoMode(huh.EchoModePassword) doesn't have any effect in accessible mode.
	form := p.newForm(
		huh.NewGroup(
			huh.NewInput().
				Title(fmt.Sprintf("Type %q to confirm deletion", requiredValue)).
				Validate(func(input string) error {
					if input != requiredValue {
						return fmt.Errorf("You entered: %q", input)
					}
					return nil
				}).
				Value(&result),
		),
	)
```

andyfeller commented on internal/prompter/speech_synthesizer_friendly_prompter_test.go:
```diff
@@ -0,0 +1,468 @@
+		expect.WithCloser(ptm, pts),
+		failOnExpectError(t),
+		failOnSendError(t),
+		expect.WithDefaultTimeout(time.Second * 600),
```
**question:** Why 10 minute / 600 second time out?

BagToad replied:
No reason - it came from other "hacceptance test" code that we were spiking. IIRC we just needed some timeout, so 🤷 

Can you recommend a different timeout instead?

williammartin replied:
Suggest making it 1 second, there shouldn't need to be much of a timeout at all since there should be no slow operations.

BagToad replied:
Fixed in ef58e627f9d2cd3698bba857d6eca1450278dbb5

andyfeller commented on internal/prompter/speech_synthesizer_friendly_prompter_test.go:
```diff
@@ -0,0 +1,525 @@
+	t.Run("InputHostname", func(t *testing.T) {
+		wg.Add(1)
+
+		hostname := "somethingdoesnotmatter.com"
```
**note:** `example.com` is a common industry-wide hostname used for testing purposes, too

BagToad replied:
```suggestion
		hostname := "example.com"
```

andyfeller commented on internal/prompter/speech_synthesizer_friendly_prompter_test.go:
```diff
@@ -0,0 +1,525 @@
+	t.Cleanup(func() { testCloser(t, console) })
+
+	t.Setenv("GH_SPEECH_SYNTHESIZER_FRIENDLY_PROMPTER", "")
+	t.Setenv("NO_COLOR", "1")
```
**question:** Why disable color for testing purposes?

BagToad replied:
hmmm I think I was having trouble getting `go-expect` to register the survey prompts and thought it might be the color, but that wasn't the issue and I forgot to remove this code. 

Removed it in fb80b5bd86cde6a009071b0b2e6d173e89abe3a9 👍

andyfeller reviewed (commented):
Solid work!  I'm gonna see about building and playing with this locally to get my hands around it. 😸 

Handful of questions I'd like to dive into around changes from the original AC and sundry points.

BagToad committed:
doc: comment typos and formatting

Co-authored-by: Andy Feller <andyfeller@github.com>

BagToad committed:
test: use example.com in tests

BagToad committed:
test(prompter): remove needless NO_COLOR set

williammartin commented on internal/prompter/prompter.go:
```diff
@@ -13,27 +15,260 @@ import (
+	// This is currently an ineffectual assignment because the value is
+	// not respected as the default in accessible mode. Leaving this in here
+	// because it may change in the future.
+	// See https://github.com/charmbracelet/huh/issues/615
```
If this is fixed in `huh` then maybe we just wait and upgrade, otherwise we could roll our own using `NewInput` that has a validator and a default.

BagToad replied:
There was race conditions accessing the terminal in the tests, causing CI to fail.

williammartin replied:
Not sure if you meant to respond to this conversation, cause I don't see the relevance.

BagToad replied:
I guess mobile messed up here 😆 Sorry

BagToad replied:
_(I hide our comments as off-topic to keep this thread clean ❤️  sorry about the confusion)_

So, I might be in favour of waiting because https://github.com/charmbracelet/huh/pull/614 is under active development, but I could be open to rolling our own.

I also want to say, I don't think this is so dramatic of a behavioral difference that it should hold up releasing these changes in preview 💭 

@andyfeller do you also have thoughts on this?

andyfeller replied:
I see `bashbunni` has reviewed the aforementioned PR with minor `golangci-lint` linter errors.

I don't see any harm if we want to mark this PR as blocked for a day while we see how that develops, catch up on other PRs.  If it stalls, then we can fall back to Will's suggestion.

Ultimately, I defer to both of you being deeper in code. 🙇

BagToad replied:
This should be fixed now 🎉

williammartin commented on internal/prompter/prompter.go:
```diff
@@ -13,27 +15,260 @@ import (
+		),
+	)
+
+	err := form.Run()
```
**Nit**

I'd prefer if we do:

```go
if err != nil {
  return "", err
}
 
return result, nil
```

BagToad replied:
Nice 🙌 

Fixed in 9cf341302eac404813eef1de4e45d8605a8985e0

williammartin commented on internal/prompter/prompter.go:
```diff
@@ -13,27 +15,260 @@ import (
+
+func (p *speechSynthesizerFriendlyPrompter) Password(prompt string) (string, error) {
+	var result string
+	// EchoMode(huh.EchoModePassword) doesn't have any effect in accessible mode.
```
Created issue here: https://github.com/charmbracelet/huh/issues/619 that may address this.

williammartin commented on internal/prompter/prompter.go:
```diff
@@ -13,27 +15,260 @@ import (
+}
+
+func (p *speechSynthesizerFriendlyPrompter) ConfirmDeletion(requiredValue string) error {
+	var result string
```
**Question**

Is there a reason to have this variable? Seems like we're never using it.

BagToad replied:
I guess not 🙂 

Removed in c4be95afd962945d123a760101e6d8b26aecd7c3

williammartin commented on internal/prompter/prompter.go:
```diff
@@ -13,27 +15,260 @@ import (
+
+func (p *speechSynthesizerFriendlyPrompter) ConfirmDeletion(requiredValue string) error {
+	var result string
+	// EchoMode(huh.EchoModePassword) doesn't have any effect in accessible mode.
```
I think this is a copy/paste error, there's no need for echo mode here.

BagToad replied:
Yeppers, fixed in 8821f77fbbe97bf3f16ad70a7ef6e1471a948da0

williammartin commented on internal/prompter/prompter.go:
```diff
@@ -13,27 +15,260 @@ import (
+		),
+	)
+
+	err := form.Run()
+	return result, err
+}
```
Same preference as above on being explicit about error and result.

BagToad replied:
As above, fixed in 9cf341302eac404813eef1de4e45d8605a8985e0

williammartin commented on internal/prompter/prompter.go:
```diff
@@ -13,27 +15,260 @@ import (
+	options := []huh.Option[string]{
+		huh.NewOption(fmt.Sprintf("Open Editor: %s", p.editorCmd), openOption),
+	}
+	if blankAllowed {
```
**Question**

Can you remind me which current flow would not allow blank? I _suspect_ we might be missing a conditional here about whether the `defaultValue` is set or not. For example, maybe you need a PR body, but it was recovered from the error/recovery file that is created to avoid losing state.

BagToad replied:
The one I tested with was `gh issue edit` and select `Body` to edit.

williammartin replied:
Ok, reading our existing code this looks ok to me.

williammartin commented on internal/prompter/speech_synthesizer_friendly_prompter_test.go:
```diff
@@ -0,0 +1,524 @@
+	var wg sync.WaitGroup
+
+	t.Run("Select", func(t *testing.T) {
+		wg.Add(1)
```
**Question**

Why do we need wait groups here? Were tests interfering with each other in some way if a `require` in a goroutine failed?

BagToad replied:
There was race conditions accessing the terminal in the tests, causing CI to fail.

BagToad replied:
Also, no I don't think it was `require` failing. 

Here's an example from failed CI https://github.com/cli/cli/actions/runs/14314579321/job/40117620523#step:5:45

BagToad replied:
I read this as tests were reading from the virtual console at the same time.

williammartin replied:
Sure looks like it though not sure why. An alternative would be to have each test create its own console. I can't see any particular advantages to sharing the console. Happy to defer though.

BagToad replied:
Is there an advantage to each test creating its own console? I figured that would make the test more cluttered with more setup.

williammartin replied:
That you don't have to have custom code in the test to avoid shared usage of a resource which should never happen in practice.

I would extract the setup to a helper, if I were to do this. Feel 2/10 strongly, don't think about it too much.

BagToad replied:
How does that work with our current hack to override `os.Stdin` - won't I still have a race condition since they all share the same IO?

williammartin replied:
Mm that's a good point, they might race on reassigning. Perhaps removable once https://github.com/charmbracelet/huh/pull/614 goes in.

BagToad replied:
Yeah. I think we'll have some follow up work to do once that goes through.

I guess let's leave it for now?

williammartin replied:
Sounds good, thanks.

BagToad replied:
Now that the IO is fixed, I've refactored the tests in 39fccf6bcf08adbe6531443eb245d282359a950c

williammartin reviewed (commented):
Lots of good stuff in here. The tests are great.

BagToad committed:
fix(prompter): use os.lookupenv for accessible prompter

BagToad committed:
test(prompter): timeout for tests is 1s

BagToad committed:
refactor(prompter): remove unused variable

BagToad committed:
doc(prompter): remove senseless comment

BagToad committed:
refactor(prompter): explicit return values

BagToad committed:
fix(prompter): rename GH_ACCESSIBLE_PROMPTER

BagToad committed:
doc(help): label GH_ACCESSIBLE_PROMPTER as preview

BagToad committed:
test(prompter):  re-add skipped test for accessible confirm default

andyfeller commented on pkg/cmd/root/help_topic.go:
```diff
@@ -109,7 +109,7 @@ var HelpTopics = []helpTopic{
 
-			%[1]sGH_SPEECH_SYNTHESIZER_FRIENDLY_PROMPTER%[1]s: set to a truthy value to enable prompts that are
+			%[1]sGH_ACCESSIBLE_PROMPTER%[1]s (preview): set to a truthy value to enable prompts that are
 			more compatible with speech synthesis based screen readers.
```
**suggest:** let's expand this a little as braille screen readers should benefit from this setting, too

```suggestion
			%[1]sGH_ACCESSIBLE_PROMPTER%[1]s (preview): set to a truthy value to enable prompts that are
			more compatible with speech synthesis and braille screen readers.
```

BagToad replied:
Sure, makes sense 👍

andyfeller commented on internal/prompter/prompter.go:
```diff
@@ -36,6 +65,218 @@ func New(editorCmd string, stdin ghPrompter.FileReader, stdout ghPrompter.FileWr
 	}
 }
 
+type speechSynthesizerFriendlyPrompter struct {
+	stdin     ghPrompter.FileReader
+	stdout    ghPrompter.FileWriter
+	stderr    ghPrompter.FileWriter
+	editorCmd string
+}
```
**question:** Do we keep the _"speech synthesizer"_ nomenclature for the type(s) and file(s) given the changes to the env var / config setting?

BagToad replied:
Good catch. I suspect not. I'll update this...

BagToad replied:
Done in 8fc8486af5ff47730c0f5021d16a6009377849d6

BagToad committed:
Merge branch 'trunk' into kw/first-pass-accessible-prompter

BagToad committed:
chore: go mod tidy

BagToad committed:
doc(prompter env): accessible prompter includes braille reader

Co-authored-by: Andy Feller <andyfeller@github.com>

BagToad committed:
refactor(prompter): rename speechSynthesizerFriendlyPrompter to accessiblePrompter

shindere commented:
Hey there,

Just to let everyone know, I am going to unsubscribe from this thread
because I am under the impression that no input from me is needed at
this stage. If you guys need me, please do not hesitate to mention me
either here or on another PR/issue in case me unsubscribing from here
makes me miss the mention. Just to clarify that if I am mentionnedhere
and do not respond it should not be interpreted as a lack of interest in
the topic, it's just because I receive quite a bunch of e-mails already.

BagToad committed:
fix(prompter): update `huh` and fix tests

BagToad committed:
fix(prompter): pass io to `huh` and refactor tests

BagToad committed:
chore: go mod tidy

williammartin commented on internal/prompter/accessible_prompter_test.go:
```diff
@@ -0,0 +1,481 @@
+	})
+}
+
+func newTestVirtualTerminal(t testing.TB) *expect.Console {
```
```
func newTestVirtualTerminal(t *testing.T) *expect.Console {
```

And other functions please.

BagToad replied:
Fixed in 47d603221d4cafebfc79619eb2296b8ea2328037

williammartin commented on internal/prompter/accessible_prompter_test.go:
```diff
@@ -0,0 +1,481 @@
+	t.Helper()
+
+	t.Setenv("GH_ACCESSIBLE_PROMPTER", "true")
+	return prompter.New("echo", console.Tty(), console.Tty(), console.Tty())
```
Think it would be useful to comment why `echo` is chosen here.

BagToad replied:
Done in 8b70870f4f553ad364795e7523c1af00714b8ccf

williammartin commented on internal/prompter/accessible_prompter_test.go:
```diff
@@ -0,0 +1,481 @@
+	"github.com/stretchr/testify/require"
+)
+
+func TestAccessiblePrompter(t *testing.T) {
```
Think a comment here might be valuable to explain that these tests are broad "wait for X, do Y, expect programmatic result". We're not testing the exact output of the prompter as we interact with it. These tests would give us value when upgrading huh, but for now what we have here gets us a lot of the way to confidence in the behaviour itself.

BagToad replied:
Done in 9eee77a2bf290558d65c6111b5c9ed53bd5648fe

williammartin commented on internal/prompter/prompter.go:
```diff
@@ -36,6 +65,213 @@ func New(editorCmd string, stdin ghPrompter.FileReader, stdout ghPrompter.FileWr
+
+	err := form.Run()
+
+	if result == "" {
+		return defaultValue, nil
+	}
```
Don't think this is needed because `result := defaultValue` is set above.

BagToad replied:
Fixed in fa0456f7ec8f78cdb4f4a286d5d2490823acc503

williammartin reviewed (approved):
Approving on the proviso that my comments are addressed as discussed on call.

BagToad committed:
test(prompter): use *testing.T instead

BagToad committed:
test(prompter): describe why echo is editorcmd

BagToad committed:
test(prompter): doc how accessible prompter tests work

BagToad committed:
fix(prompter): remove needless default value assignment

andyfeller commented:
## Thoughts

1. `gh` prompts seem to use parenthesis in a couple different ways, unsure if confusing to others
   - `(required)` for required fields
   - `(<VALUE>)` for default values

1. `huh` displays empty default value, could probably avoid showing this

   ```shell
   Description ()
    
   Input:
   ```

1. `huh` has some inconsistent spacing around prompt and input

   - selects and multi-selects have no spacing after prompt
    
     ```shell
     What would you like to do?
     1. Create a new repository on GitHub from scratch
     3. Create a new repository on GitHub from a template repository
     4. Push an existing local repository to GitHub
    
     Choose: 
     ```
    
   - inputs have spacing after prompt
    
     ```shell
     Path to local repository (.)
    
     Input:
     ```

1. `huh` validation error message might not be instructive enough for screen readers

   ```shell
   gh issue delete 151
   ! Deleted issues cannot be recovered.
   Type "151" to confirm deletion
    
   Input: 150
   You entered: "150"
   Input: 152
   You entered: "152"
   Input: 
   You entered: ""
   Input: ^C 
   ```
    
   versus current experience:
    
   ```shell
   gh issue delete 151
   ! Deleted issues cannot be recovered.
   X Sorry, your reply was invalid: You entered 150
   ? Type 151 to confirm deletion: 
   ```

1. Markdown editor prompt shows an extra colon in the editor option, where is this coming from?

   ```shell
   Body
   1. Open Editor: 
   3. Skip
   ```

## Demo

### `Prompter.AuthToken`

- Accessible experience

  <img width="1512" alt="gh-auth-login-accessible" src="https://github.com/user-attachments/assets/efa95541-4989-4506-852b-a7dc6ccc4fc4" />

  ```shell
  $ gh auth login
  Where do you use GitHub?
  1. GitHub.com
  4. Other
  
  Choose: 1
  Chose: GitHub.com
                   
  What is your preferred protocol for Git operations on this host?
  1. HTTPS
  4. SSH
  
  Choose: 1
  Chose: HTTPS
              
  How would you like to authenticate GitHub CLI?
  1. Login with a web browser
  2. Paste an authentication token
  
  Choose: 2
  Chose: Paste an authentication token
                                      
  Tip: you can generate a Personal Access Token here https://github.com/settings/tokens
  The minimum required scopes are 'repo', 'read:org', 'workflow'.
  Paste your authentication token:
  
  Input: ghp_OBSCURED
  Input: ghp_OBSCURED
                                                 
  - gh config set -h github.com git_protocol https
  ✓ Configured git protocol
  ✓ Logged in as andyfeller
  ! You were already logged in to this account
  ```

- Default experience

  <img width="1512" alt="gh-auth-login-default" src="https://github.com/user-attachments/assets/8cf2c347-1dd4-49c4-858a-7d0106ed2e3b" />

  ```shell
  $ gh auth login
  ? Where do you use GitHub? GitHub.com
  ? What is your preferred protocol for Git operations on this host? HTTPS
  ? How would you like to authenticate GitHub CLI? Paste an authentication token
  Tip: you can generate a Personal Access Token here https://github.com/settings/tokens
  The minimum required scopes are 'repo', 'read:org', 'workflow'.
  ? Paste your authentication token: ****************************************
  - gh config set -h github.com git_protocol https
  ✓ Configured git protocol
  ✓ Logged in as andyfeller
  ! You were already logged in to this account
  ```

### `Prompter.Confirm`

- Accessible experience

  <img width="1512" alt="gh-repo-create-accessible-2" src="https://github.com/user-attachments/assets/563ffc9e-aea5-40f3-9c5b-ccff482f09cc" />

  ```shell
  Add a remote?
  
  Choose [Y/n]: 
  Chose: Yes
  ```

- Default experience

  <img width="1512" alt="gh-repo-create-default" src="https://github.com/user-attachments/assets/879f435f-b2f6-4df1-89db-500d80c8f027" />

  ```shell
  ? Add a remote? (Y/n) 
  ```

### `Prompter.ConfirmDeletion`

- Accessible experience

  <img width="1512" alt="gh-repo-delete-accessible" src="https://github.com/user-attachments/assets/b597f023-2275-4cea-be13-6128f1dcc457" />

  ```shell
  $ gh repo delete
  Type "tinyfists/gh-demo-accessible" to confirm deletion
  
  Input: tinyfists/gh-demo-accessible
  Input: tinyfists/gh-demo-accessible
                                     
  ✓ Deleted repository tinyfists/gh-demo-accessible
  ```

- Default experience

  <img width="1512" alt="gh-repo-delete-default" src="https://github.com/user-attachments/assets/685a1c70-da89-49a6-9017-bce07047cbbb" />

  ```shell
  $ gh repo delete    
  ? Type tinyfists/gh-demo-default to confirm deletion: tinyfists/gh-demo-default
  ✓ Deleted repository tinyfists/gh-demo-default
  ```

### `Prompter.InputHostname`

- Accessible experience

  <img width="1512" alt="gh-auth-login-host-accessible" src="https://github.com/user-attachments/assets/51e80264-0233-4d66-baa4-9fe05a94196a" />

   ```shell
   gh auth login
   Where do you use GitHub?
   1. GitHub.com
   2. Other
    
   Choose: 2
   Chose: Other
                
   Hostname:
    
   Input: github.com
   Input: github.com
   ```

- Default experience

  <img width="1512" alt="gh-auth-login-host-default" src="https://github.com/user-attachments/assets/03046120-dd80-4054-bb48-c0d6ccf6073e" />

   ```shell
   gh auth login
   ? Where do you use GitHub? Other
   ? Hostname: github.com
   ? What is your preferred protocol for Git operations on this host?  [Use arrows to move, type to filter]
   > HTTPS
     SSH
   ```

### `Prompter.MarkdownEditor`

- Accessible experience

  <img width="1512" alt="gh-issue-create-accessible-1" src="https://github.com/user-attachments/assets/833c9fdd-246d-4b40-9d77-bd4d4b4b51ff" />
  
  ```shell
  $ gh issue create
  
  Creating issue in tinyfists/issue-driven-github-admin
  
  Title (required) ()
  
  Input: Demonstration of accessible issue create
  Input: Demonstration of accessible issue create
                                                 
  Body
  1. Open Editor: 
  2. Skip
  
  Choose: 1
  Chose: Open Editor: 
                      
  What's next?
  1. Submit
  2. Continue in browser
  5. Add metadata
  6. Cancel
  ```

- Default experience

  <img width="1512" alt="gh-issue-create-default-1" src="https://github.com/user-attachments/assets/ddbcfd47-6bc2-44fd-9e80-827eba25987b" />

  ```shell
  $ gh issue create
  
  Creating issue in tinyfists/issue-driven-github-admin
  
  ? Title (required) Demo default issue create
  ? Body [(e) to launch vim, enter to skip] 
  ```

### `Prompter.MultiSelect`

- Accessible experience

  <img width="1512" alt="gh-issue-create-accessible-multiselect" src="https://github.com/user-attachments/assets/0c2a09c3-c589-4fc6-9072-bb5a2120c471" />

  ```shell
  What would you like to add?
  1.   Assignees
  2.   Labels
  3.   Projects
  4.   Milestone
  
  Select up to 4 options. 0 to continue.
  Select: 1
  Selected: Assignees
  
  What would you like to add?
  1. ✓ Assignees
  2.   Labels
  3.   Projects
  4.   Milestone
  
  Select up to 4 options. 0 to continue.
  Select: 2
  Selected: Labels
  
  What would you like to add?
  1. ✓ Assignees
  2. ✓ Labels
  3.   Projects
  4.   Milestone
  
  Select up to 4 options. 0 to continue.
  Select: 0
  Selected: Assignees, Labels
  ```

- Default experience

  <img width="1512" alt="gh-issue-create-default-multiselect" src="https://github.com/user-attachments/assets/962c017c-b98a-48df-b3d9-2ee9f0b743d9" />
  
  ```shell
  $ gh issue create
  
  Creating issue in tinyfists/issue-driven-github-admin
  
  ? Title (required) Demo default issue create
  ? Body <Received>
  ? What's next? Add metadata
  ? What would you like to add?  [Use arrows to move, space to select, <right> to all, <left> to none, type to filter]
    [x]  Assignees
  > [x]  Labels
    [ ]  Projects
    [ ]  Milestone
  ```

### `Prompter.Select`

- Accessible experience

  <img width="1512" alt="gh-issue-create-accessible-1" src="https://github.com/user-attachments/assets/33d2c588-3400-48ee-831a-c1ce85bc90b1" />

  ```shell
  What's next?
  1. Submit
  2. Continue in browser
  3. Add metadata
  4. Cancel
  
  Choose: 3
  Chose: Add metadata
  ```

- Default experience

  <img width="1512" alt="gh-issue-create-default-1" src="https://github.com/user-attachments/assets/14123e5d-55bf-460b-b00e-da5172946e4b" />

  ```shell
  ? What's next?  [Use arrows to move, type to filter]
  > Submit
    Continue in browser
    Add metadata
    Cancel
  ```
  
  ```shell
  ? What's next? Add metadata
  ? What would you like to add?  [Use arrows to move, space to select, <right> to all, <left> to none, type to filter]
  > [ ]  Assignees
    [ ]  Labels
    [ ]  Projects
    [ ]  Milestone
  ```

andyfeller commented on internal/prompter/prompter.go:
```diff
@@ -36,6 +65,209 @@ func New(editorCmd string, stdin ghPrompter.FileReader, stdout ghPrompter.FileWr
+	var result string
+	skipOption := "skip"
+	openOption := "open"
+	options := []huh.Option[string]{
+		huh.NewOption(fmt.Sprintf("Open Editor: %s", p.editorCmd), openOption),
+	}
```
I don't know if `p.editorCmd` always has a value, which results in the `Open Editor` having no editor listed:

```shell
$ echo $EDITOR
vim

$ gh config get editor

$ gh issue create

Creating issue in tinyfists/issue-driven-github-admin

Title (required) ()

Input: This is just a title
Input: This is just a title
                           
Body
1. Open Editor: 
2. Skip

Choose: 
```

versus the current experience:

```shell
$ gh issue create

Creating issue in tinyfists/issue-driven-github-admin

? Title (required) Demo default issue create
? Body [(e) to launch vim, enter to skip] 
```

### How does this work today?

Currently, `gh` has relied upon `go-survey` to prompt the user whether to open the editor using the editor we have from either 1) configuration or 2) env var:

https://github.com/cli/cli/blob/057ad52fdb05032b8501f3dc2fc5dc174c4a86ae/pkg/surveyext/editor_manual.go#L43-L74

https://github.com/cli/cli/blob/057ad52fdb05032b8501f3dc2fc5dc174c4a86ae/pkg/surveyext/editor.go#L22-L32

In the code above, `gh` is having to ask the user if they want to open the editor, however the logic for figuring out the editor is buried in the `surveyext` package after this code stanza.

### What do we do to address this?

1. Change the user experience by removing the editor name from the prompt
2. Duplicate the logic for using the `defaultEditor` if the editor command is empty
3. Refactor the logic for the default editor so other places can use it

BagToad replied:
Fixed in b8cd094ca8f7a3cf3ab77a1e69f94eaae6403626

andyfeller committed:
Ensure markdown confirm prompt shows editor name

Apparently, `gh` might not actually have an editor at the time we're prompting
the user if they want to use it for markdown editing.  In the survey package,
there is a function that will handle fallback to the default editor based on
environment variables and parse it in the case the editor contains flags and
arguments for cases like Visual Studio Code.

Additionally, there are no tests for the EditorName function and the fact it
is loaded via `init` makes this difficult to test.

Co-authored-by: Kynan Ware <47394200+BagToad@users.noreply.github.com>

andyfeller reviewed (approved):
An astounding heavy lift to get this prompting experience! ✨ 

I think there are things we should follow up after this PR including additional UX testing in Linux and Windows, working with our friends at Charm, refactoring some logic to be testable, etc.  I don't think there is anything to block this PR but we do have more work beyond this.

BagToad committed:
test(prompter):  fix race condition

This test was trying to block on `expect`’ing a string at the same
time the prompt was completed.

This doesn't need to happen for this test. It should just check for
the output from the Input prompt invocation.

BagToad commented on internal/prompter/prompter.go:
```diff
@@ -36,6 +65,209 @@ func New(editorCmd string, stdin ghPrompter.FileReader, stdout ghPrompter.FileWr
+		return "", nil
+	}
+
+	// openOption was selected
```
```suggestion
	// launchOption was selected
```

BagToad committed:
test(prompter): fix invalid comment
